ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -629,6 +646,165 @@ private IndexValue findKey(StoreKey key, FileSpan fileSpan, EnumSet<IndexEntryTy
+              + " Undelete LifeVersion: " + lifeVersion, StoreErrorCodes.Life_Version_Conflict);
+    }
+  }
+
+  /**
+   * Ensure that the previous {@link IndexValue}s is structured correctly for undeleting the {@code key} when there is
+   * no lifeVersion provided.
+   * <p/>
+   * Undelete should be permitted only when the first record is a Put and last record is a Delete, and the Put record
+   * hasn't expired yet.
+   * @param key the key to be undeleted.
+   * @param values the previous {@link IndexValue}s in reversed order.
+   */
+  void validateSanityForUndeleteWithoutLifeVersion(StoreKey key, List<IndexValue> values) throws StoreException {
+    // When it's valid to undelete this key
+    // P/T + D
+    // P/T + D + U + D
+    if (values.size() == 1) {
+      IndexValue value = values.get(0);
+      if (value.isDelete() || value.isTTLUpdate()) {
+        throw new StoreException("Id " + key + " is compacted in index" + dataDir,
+            StoreErrorCodes.ID_Deleted_Permanently);
+      } else if (value.isPut()) {
+        throw new StoreException("Id " + key + " is not deleted yet in index " + dataDir,
+            StoreErrorCodes.ID_Not_Deleted);
+      } else {
+        throw new StoreException("Id " + key + " is already undeleted in index" + dataDir,
+            StoreErrorCodes.ID_Undeleted);
+      }
+    }
+    // First item has to be put and last item has to be a delete.
+    // PutRecord can't expire and delete record can't be older than the delete retention time.
+    IndexValue firstValue = values.get(values.size() - 1);
+    IndexValue lastValue = values.get(0);
+    if (lastValue.isUndelete()) {
+      throw new StoreException("Id " + key + " is already undeleted in index" + dataDir, StoreErrorCodes.ID_Undeleted);
+    }
+    if (!firstValue.isPut() || !lastValue.isDelete()) {
+      throw new StoreException(
+          "Id " + key + " requires first value to be a put and last value to be a delete in index " + dataDir,
+          StoreErrorCodes.ID_Not_Deleted);
+    }
+    if (lastValue.getOperationTimeInMs() + TimeUnit.DAYS.toMillis(config.storeDeletedMessageRetentionDays)
+        < time.milliseconds()) {
+      throw new StoreException("Id " + key + " already permanently deleted in index " + dataDir,
+          StoreErrorCodes.ID_Deleted_Permanently);
+    }
+    maybeChangeExpirationDate(firstValue, values);
+    if (isExpired(firstValue)) {
+      throw new StoreException("Id " + key + " already expired in index " + dataDir, StoreErrorCodes.TTL_Expired);
+    }
+  }
+
+  /**
+   * Change the target's expiration date and set the ttl_update_index to be true if there is a ttl index in given list.
+   * @param target the {@link IndexValue} to change expiration date.
+   * @param allValues the given list of {@link IndexValue}s.
+   */
+  void maybeChangeExpirationDate(IndexValue target, List<IndexValue> allValues) {
+    if (target.isTTLUpdate()) {
+      return;
```

Comment:
if target is already updated or expiration time = -1, in either case we can skip update